sycl/doc/DeviceGlobal.md

```diff
@@ -110,60 +101,137 @@ However, when the property is present, it has a member variable which is the
 type itself, and member functions return a reference to this value.
 
 ```
-template<typename T>
+template<typename T, typename PropertyListT>
 class device_global {
   T val;
  public:
+  device_global() : val() {}
   T& get() noexcept { return val; }
   /* other member functions */
 };
 ```
 
+Note that the default constructor "value initializes" the `val` member
+variable.  Since the type `T` is limited to types that are trivially
+constructible, this means that `val` will be zero initialized.
+
 In both cases the member variable (either `usmptr` or `val`) must be the first
 member variable in the class.  As we will see later, the runtime assumes that
 the address of the `device_global` variable itself is the same as the address
 of this member variable.
 
+#### Attributes attached to the class
+
+The `device_global` class declaration contains three C++ attributes which
+convey information to the front-end.  These attributes are only needed for the
+device compiler, and the `#ifdef __SYCL_DEVICE_ONLY__` allows the customer to
+use another host compiler, even if it does not recognize these attributes.
+Also note that these attributes are all in the `__sycl_detail__` namespace, so
+they are considered implementation details of DPC++.  We do not intend to
+support them as general attributes that customer code can use.
+
+```
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename T, typename ...Props>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    "sycl-device-global-size",
+    Props::meta_name...,
+    sizeof(T),
+    Props::meta_value...
+    )]]
+  [[__sycl_detail__::global_variable_allowed]]
+  [[__sycl_detail__::restrictions(device_global)]]
+#endif
+  device_global<T, property_list<Props...>> {/*...*/};
+```
+
+The `[[__sycl_detail__::add_ir_global_variable_attributes()]]` attribute is
+used to convey information about the compile-time properties to the front-end,
+and it is described more fully by the [compile-time properties][2] design
+document.  This attribute is also used for other classes that have properties,
+so it is not specific to the `device_global` class.
+
+[2]: <https://github.com/intel/llvm/pull/4937>
+
+Note that the parameter list to
+`[[__sycl_detail__::add_ir_global_variable_attributes()]]` contains one
+additional property named `"sycl-device-global-size"`.  The `sycl-post-link`
```

Comment:
Could you please clarify if there is a specific reason why the size needs to be passed explicitly?

Reply:
The `sycl-post-link` tool needs to know the size because it writes it out as part of the "SYCL/device globals" property set.  I thought it would be easier for the `sycl-post-link` tool to get the size via an IR attribute like this rather than requiring the `sycl-post-link` tool to parse through the template parameters to figure it out.

Reply:
Why `sycl-post-link` can't just take `sizeof(global_variable)`?

Reply:
Two reasons:

1. The `sizeof(global_variable)` is not the same as `sizeof(T)` in the case when the variable does not have the `device_image_scope` property.  In this case the variable just contains a pointer, so the size is 8 bytes regardless of the type of `T`.

2. Since `device_global` is a template class, the template parameters need to be known at compile time in order to specialize the template class for a particular `T`.  However, the user provides the type `T` in their application, but the `sycl-post-link` tool is provided as a pre-compiled binary to the user.  Unless the user recompiles `sycl-post-link` from source, I don't see how this could work.